### DISCUSSION TITLE
Cookie & CORS

### DISCUSSION THREAD

**Discussion by klaydev** [fastapi-users/fastapi-users#1174] [Q&A] [answered]
Hello everyone. I turned on CARS and now authentication with cookies does not work, I did everything according to the fastape and fastapi-users documentation

**Comment by frankie567** [comment]
CORS define how a browser is allowed to make **cross-origin requests**.

Typically, if you have a frontend that is hosted on `app.mydomain.com` and a backend hosted on `api.mydomain.com`, the browser will try to make requests from the frontend to the backend.

```mermaid
flowchart LR
    A[app.mydomain.com]
    B[api.mydomain.com]
    A --> B
```

Problem is, there are **not on the same domain**. For security reasons, browsers don't allow JavaScript to make requests to a site on another domain. Besides, they usually block sensitive headers like cookies.

That's when CORS come into action: it's a way to tell the browser to **relax** this security measure for some site and circumstances.

In the example above, we need to configure CORS so that:

* `app.mydomain.com` can make requests to the backend
* The authentication cookie is not removed from the request

With the FastAPI/Starlette middleware, it would look like this:

```py
app.add_middleware(
    CORSMiddleware,
    allow_origins=["app.mydomain.com"],   # All the *frontend* domains that are allowed to make requests 
    allow_credentials=True,  # Allow cookies to be part of the request
    allow_methods=["*"],
    allow_headers=["*"],
)
```

**Reply by klaydev** [reply]
Thanks. But the problem is that the frontend and backend are developed by two different people who keep them at a local address. I tried adding their addresses to CARS, the code 200 OK comes, but cookies are not transmitted

**Reply by frankie567** [reply, marked as answer]
The local address of the frontend should be added to `allow_origins` indeed. `allow_credentials` should be set to `True` for the cookies to be transmitted.

---

Be careful also with the [`cookie_secure` flag of `CookieTransport`](https://fastapi-users.github.io/fastapi-users/10.4/configuration/authentication/transports/cookie/#configuration). If you're not over HTTPS/SSL, the flag should be set to `False`.

---

In the frontend, if `fetch` is used to make requests, credentials should explicitly be allowed: https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Using_Fetch#sending_a_request_with_credentials_included

**Reply by klaydev** [reply]
Thanks!
